File: exercise/kotlin/day07/src/main/kotlin/ci/Pipeline.kt
package ci

class Pipeline(
    private val config: Config,
    private val emailSender: Emailer,
    private val log: Logger
) {
    fun run(project: Project) {
        var testsPassed: Boolean
        val deploySuccessful: Boolean

        testsPassed = if (project.hasTests()) {
            if ("success" == project.runTests()) {
                log.info("Tests passed")
                true
            } else {
                log.error("Tests failed")
                false
            }
        } else {
            log.info("No tests")
            true
        }
        deploySuccessful = if (testsPassed) {
            if ("success" == project.deploy()) {
                log.info("Deployment successful")
                true
            } else {
                log.error("Deployment failed")
                false
            }
        } else {
            false
        }
        if (config.sendEmailSummary()) {
            log.info("Sending email")
            if (testsPassed) {
                if (deploySuccessful) {
                    emailSender("Deployment completed successfully")
                } else {
                    emailSender("Deployment failed")
                }
            } else {
                emailSender("Tests failed")
            }
        } else {
            log.info("Email disabled")
        }
    }
}

fun interface Config {
    fun sendEmailSummary(): Boolean
}

typealias Emailer = (message: String) -> Unit

interface Logger {
    fun info(message: String)
    fun error(message: String)
}

class Project private constructor(private val buildsSuccessfully: Boolean, private val testStatus: TestStatus) {
    fun hasTests(): Boolean = testStatus !== TestStatus.NO_TESTS
    fun runTests(): String = if (testStatus === TestStatus.PASSING_TESTS) "success" else "failure"
    fun deploy(): String = if (buildsSuccessfully) "success" else "failure"

    class ProjectBuilder {
        private var buildsSuccessfully = false
        private var testStatus: TestStatus? = null
        fun setTestStatus(testStatus: TestStatus?): ProjectBuilder {
            this.testStatus = testStatus
            return this
        }

        fun setDeploysSuccessfully(buildsSuccessfully: Boolean): ProjectBuilder {
            this.buildsSuccessfully = buildsSuccessfully
            return this
        }

        fun build(): Project = Project(buildsSuccessfully, testStatus!!)
    }
}

enum class TestStatus {
    NO_TESTS,
    PASSING_TESTS,
    FAILING_TESTS
}

File: exercise/ts/day07/src/pipeline.ts
import {Config, Emailer, Logger} from "./ci";
import {Project} from "./project";

export class Pipeline {
    private readonly config: Config;
    private readonly emailer: Emailer;
    private readonly log: Logger;

    constructor(config: Config, emailer: Emailer, log: Logger) {
        this.config = config;
        this.emailer = emailer;
        this.log = log;
    }

    run(project: Project): void {
        let testsPassed: boolean;
        let deploySuccessful: boolean;

        if (project.hasTests()) {
            if ('success' === project.runTests()) {
                this.log.info('Tests passed');
                testsPassed = true;
            } else {
                this.log.error('Tests failed');
                testsPassed = false;
            }
        } else {
            this.log.info('No tests');
            testsPassed = true;
        }

        if (testsPassed) {
            if ('success' === project.deploy()) {
                this.log.info('Deployment successful');
                deploySuccessful = true;
            } else {
                this.log.error('Deployment failed');
                deploySuccessful = false;
            }
        } else {
            deploySuccessful = false;
        }

        if (this.config.sendEmailSummary()) {
            this.log.info('Sending email');
            if (testsPassed) {
                if (deploySuccessful) {
                    this.emailer.send('Deployment completed successfully');
                } else {
                    this.emailer.send('Deployment failed');
                }
            } else {
                this.emailer.send('Tests failed');
            }
        } else {
            this.log.info('Email disabled');
        }
    }
}